ch2.2.1-ex2/permutation-to-operations.ts

```diff
@@ -0,0 +1,83 @@
+import { Operation } from "./operations";
+
+export class NotAPermutation extends Error {}
+
+export class CannotOutputInOriginalOrder extends Error {
+    public firstNumber: number;
+    public secondNumber: number;
+    public previousHighNumber: number;
+
+    public constructor(
+        firstNumber: number,
+        secondNumber: number,
+        previousHighNumber: number,
+    ) {
+        super();
+        this.firstNumber = firstNumber;
+        this.secondNumber = secondNumber;
+        this.previousHighNumber = previousHighNumber;
+    }
+
+    public toString(): string {
+        return `Can't output ${this.firstNumber} then ${
+            this.secondNumber
+        }; the former was stacked in order to output ${
+            this.previousHighNumber
+        }`;
+    }
+}
+
+// Returns `true` iff the array is some permutation of [1, 2, ..., n], n >= 1
+function isPermutation(array: number[]): boolean {
+    return (
+        array.length > 0 &&
+        array.every((_, i): boolean => array.includes(i + 1))
+    );
```

Comment:
I'm a little bit happy with this function. I don't believe I've used `.every` to count indices before.

The `.include` makes it an `O(n^2)` algorithm. A `.sort` (on a shallow-cloned array) plus a `.every` would've worked out faster, at least in the average case. Using a [`TreeSet`](https://docs.oracle.com/javase/8/docs/api/?java/util/TreeSet.html) (which JavaScript doesn't have naively, so we'll have to implement or import), one can reach `O(n log n)` worst-case performance.

But using those more complicated algos to gain a speedup is most likely premature optimization - the permutation syntax limits the arrays to no more than 9 elements. A 9*9 nested loop is most likely fine. So simpler code it is.